adapters/richaudience/richaudience.go

```diff
@@ -0,0 +1,355 @@
+				Bid:     &bid,
+				BidType: "banner",
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+
+	return bidResponse, nil
+}
+
+func setHeaders(raiHeaders *http.Header) {
+	raiHeaders.Set("Content-Type", "application/json;charset=utf-8")
+	raiHeaders.Set("Accept", "application/json")
+	raiHeaders.Add("X-Openrtb-Version", "2.5")
+}
+
+func setImp(request *openrtb2.BidRequest, richaudienceRequest *richaudienceRequest) (resImps []openrtb2.Imp, err error) {
+	//Imp: Id, tagId, Secure, Bidfloor, Bidfloorcur, Banner, test
+	//Banner: Id, W, H, Format
+	//Format: W, H
+
+	for _, imp := range request.Imp {
+		var secure = int8(0)
+		raiExt, errImp := parseImpExt(&imp)
+
+		if errImp != nil {
+			return nil, errImp
+		}
+
+		if raiExt != nil {
+			if raiExt.Pid != "" {
+				imp.TagID = raiExt.Pid
+			}
+
+			if raiExt.TestRa {
+				richaudienceRequest.Test = int8(1)
+			}
+
+			if raiExt.BidFloor <= 0 {
+				imp.BidFloor = 0.00001
+			} else {
+				imp.BidFloor = raiExt.BidFloor
+			}
+
+			if raiExt.BidFloorCur == "" {
+				imp.BidFloorCur = "USD"
+			} else {
+				imp.BidFloorCur = raiExt.BidFloorCur
+			}
+		}
+
+		if request.Site != nil && request.Site.Page != "" {
+			pageURL, error := url.Parse(request.Site.Page)
+			if error == nil && pageURL.Scheme == "https" {
+				secure = int8(1)
+			}
+		}
+		imp.Secure = &secure
+
+		if imp.Banner == nil {
+			err = &errortypes.BadInput{
+				Message: "Banner Object not found",
+			}
+			return nil, err
+		} else {
+			raiBanner := *imp.Banner
+			if raiBanner.W == nil && raiBanner.H == nil {
+				if len(raiBanner.Format) == 0 {
+					err = &errortypes.BadInput{
+						Message: "Format Object not found",
+					}
+					return nil, err
+				} else {
+					imp.Banner = &raiBanner
+				}
+			}
+		}
+
+		resImps = append(resImps, imp)
+
+	}
+	return resImps, nil
+}
+
+func setSite(request *openrtb2.BidRequest, richaudienceRequest *richaudienceRequest) {
+	if request.Site != nil {
+		if request.Site.Page != "" {
+			if request.Site.Domain == "" {
+				pageURL, error := url.Parse(request.Site.Page)
+				if error == nil {
+					richaudienceRequest.Site.Domain = pageURL.Host
+				}
+			} else {
+				richaudienceRequest.Site.Domain = request.Site.Domain
+			}
+
+			richaudienceRequest.Site.Page = request.Site.Page
+		}
+	}
+}
+
+func setDevice(request *openrtb2.BidRequest, richaudienceRequest *richaudienceRequest) (err error) {
+
+	if request.Device.IP == "" {
+		err = &errortypes.BadInput{
+			Message: "request.Device.IP is required",
+		}
+		return err
+	} else {
+		richaudienceRequest.Device.IP = request.Device.IP
+	}
+
+	if request.Device.DNT != nil {
+		richaudienceRequest.Device.DNT = *request.Device.DNT
+	} else {
+		richaudienceRequest.Device.DNT = 0
+	}
+
+	if request.Device.Lmt != nil {
+		richaudienceRequest.Device.Lmt = *request.Device.Lmt
+	} else {
+		richaudienceRequest.Device.Lmt = 0
+	}
+
+	if request.Device.UA != "" {
+		richaudienceRequest.Device.UA = request.Device.UA
+	}
```

Comment:
You don't need the check for an empty string here. A direct assignment will yield the same result.

Reply:
in next commit i have solved